File: Backend/src/main/java/com/example/medisphere/service/PrescriptionService.java
package com.example.medisphere.service;

import com.example.medisphere.model.Prescription;
import com.example.medisphere.repository.PrescriptionRepository;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class PrescriptionService {

    private final PrescriptionRepository prescriptionRepository;

    public PrescriptionService(
            PrescriptionRepository prescriptionRepository) {

        this.prescriptionRepository = prescriptionRepository;
    }

    // Create prescription
    public Prescription addPrescription(
            Prescription prescription) {

        return prescriptionRepository.save(prescription);
    }

    // Get all prescriptions
    public List<Prescription> getAllPrescriptions() {

        return prescriptionRepository.findAll();
    }

    // Get prescriptions by patient ID
    public List<Prescription> getPrescriptionsByPatientId(
            String patientId) {

        return prescriptionRepository.findByPatientId(
                patientId
        );
    }

    // Get prescription by ID
    public Optional<Prescription> getPrescriptionById(
            String id) {

        return prescriptionRepository.findById(id);
    }

    // Update prescription
    public Prescription updatePrescription(
            String id,
            Prescription updatedPrescription) {

        return prescriptionRepository.findById(id)
                .map(prescription -> {

                    prescription.setPatientId(
                            updatedPrescription.getPatientId()
                    );

                    prescription.setDoctorId(
                            updatedPrescription.getDoctorId()
                    );

                    prescription.setAppointmentId(
                            updatedPrescription.getAppointmentId()
                    );

                    prescription.setMedicineName(
                            updatedPrescription.getMedicineName()
                    );

                    prescription.setDosage(
                            updatedPrescription.getDosage()
                    );

                    prescription.setInstructions(
                            updatedPrescription.getInstructions()
                    );

                    return prescriptionRepository.save(
                            prescription
                    );
                })
                .orElse(null);
    }

    // Delete prescription
    public boolean deletePrescription(String id) {

        if (prescriptionRepository.existsById(id)) {

            prescriptionRepository.deleteById(id);

            return true;
        }

        return false;
    }
}

File: Frontend/src/services/prescriptionService.js
const API_URL = "http://localhost:8080/api/prescriptions";

// Get all prescriptions
export const getPrescriptions = async () => {
  const response = await fetch(API_URL);

  if (!response.ok) {
    throw new Error("Failed to fetch prescriptions");
  }

  return response.json();
};

// Get prescriptions by patient ID
export const getPrescriptionsByPatientId = async (patientId) => {
  const response = await fetch(`${API_URL}/patient/${patientId}`);

  if (!response.ok) {
    throw new Error("Failed to fetch patient prescriptions");
  }

  return response.json();
};

// Get prescription by ID
export const getPrescriptionById = async (id) => {
  const response = await fetch(`${API_URL}/${id}`);

  if (!response.ok) {
    throw new Error("Failed to fetch prescription");
  }

  return response.json();
};

// Create prescription
export const createPrescription = async (prescription) => {
  const response = await fetch(API_URL, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(prescription),
  });

  if (!response.ok) {
    throw new Error("Failed to create prescription");
  }

  return response.json();
};

// Update prescription
export const updatePrescription = async (id, prescription) => {
  const response = await fetch(`${API_URL}/${id}`, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(prescription),
  });

  if (!response.ok) {
    throw new Error("Failed to update prescription");
  }

  return response.json();
};

// Delete prescription
export const deletePrescription = async (id) => {
  const response = await fetch(`${API_URL}/${id}`, {
    method: "DELETE",
  });

  if (!response.ok) {
    throw new Error("Failed to delete prescription");
  }

  return true;
};